server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -2238,12 +2238,10 @@ private void deleteSnapshotsFromRepository(
             assert deleteEntry.state() == SnapshotDeletionsInProgress.State.STARTED : "incorrect state for entry [" + deleteEntry + "]";
             final Repository repository = repositoriesService.repository(deleteEntry.repository());
 
-            // TODO: Relying on repository flag to decide delete flow may lead to shallow snapshot blobs not being taken up for cleanup
-            // when the repository currently have the flag disabled and we try to delete the shallow snapshots taken prior to disabling
-            // the flag. This can be improved by having the info whether there ever were any shallow snapshot present in this repository
-            // or not in RepositoryData.
-            // SEE https://github.com/opensearch-project/OpenSearch/issues/8610
-            final boolean cleanupRemoteStoreLockFiles = REMOTE_STORE_INDEX_SHALLOW_COPY.get(repository.getMetadata().settings());
+            // If there is at least one shallow copy snapshot in repository, we can use the new method
+            // deleteSnapshotsAndReleaseLockFiles which will release lock files as well during snapshot deletion.
+            boolean cleanupRemoteStoreLockFiles = snapshotIds.stream()
+                .anyMatch(snapshotId -> repositoryData.getSnapshotType(snapshotId) == SnapshotType.SHALLOW_COPY);
```

Comment:
can we use `SnapshotInfo` instead to determine if snapshot is a shallow one  ?

Reply:
problem with that is we we have to additionally read one file for each snapshot in that case.

Reply:
Right , looks like it is not that expensive as well . The benefit of this approach is that this will work for existing clusters as well.

Reply:
this approach will also work on existing clusters as we will be backfilling this data on existing `RepositoryData` as well as part of the first `RepositoryData` update once this change is deployed.